aten/src/ATen/native/cuda/jit_utils.cpp

```diff
@@ -933,11 +933,55 @@ void initializeCudaContext() {
   }
 }
 
+#ifdef USE_ROCM
+int calc_io_size(
+    const int nInputs,
+    const int nOutputs,
+    const c10::ScalarType& inputs_type,
+    const c10::ScalarType& result_type) {
+    if (nInputs > 0 && nOutputs > 0) {
+        return std::min(c10::elementSize(inputs_type), c10::elementSize(result_type));
+    }
+
+    if (nInputs > 0) {
+        return c10::elementSize(inputs_type);
+    }
+
+    if (nOutputs > 0) {
+        return c10::elementSize(result_type);
+    }
+
+    return 0;
+}
+#endif
+
+int calc_thread_work_size(
```

Comment:
Is this introducing a new function that is only used on ROCM?

Reply:
yes

Reply:
calc_io_size() is ROCM only, which is used in the below function calc_thread_work_size().
For non_ROCM, calc_thread_work_size() defaults to 4 (same as before)

Reply:
Could we avoid defining a function for both if it is only used in one path?

Reply:
@eqy 
We need a bare function for non-ROCM as well for calc_thread_work_size(), to return the default value of 4.
This is because the JIT Loop code template (in jit_utils.cpp) now takes thread_work_size as a parameter instead of being hardcoded.

So we need to determine the thread_work_size in CUDJitLoop.cuh & jiterator.cu, and then pass it to the at::cuda::jit::generate_code() for template substitution

Reply:
We could avoid calling this function for non-ROCM; but then the call-sites in CUDJitLoop.cuh & jiterator.cu will need to carry the #ifdef for ROCM vs non_ROCM. Since this run in host code there shudn't be any perf concerns